htlcswitch/decayedlog.go

```diff
@@ -253,6 +297,191 @@ func (d *DecayedLog) gcExpiredHashes(height uint32) (uint32, error) {
+	// Find them, so that we can prepare them for garbage collection.
+	batchReplay := tx.ReadBucket(batchReplayBucket)
+	setExpiry := tx.ReadBucket(replaySetExpiryBucket)
+	var setsWithoutExpiry [][]byte
+	numSetsWithoutExpiry := 0
+
+	// As the bucket may contain several hundred million entries, we cannot
+	// just load those IDs into memory at once. Instead, we scan
+	// sequentially and abort after MaxNumberOfSetsRemovedPerBlock entries
+	// without expiry have been found. This way at least a fraction of
+	// all old entries will be garbage collected while keeping resource
+	// usage limited.
+	breakForEachPseudoError := errors.New("break out of ForEach")
+	err := batchReplay.ForEach(func(id, v []byte) error {
+		if numSetsWithoutExpiry >= MaxNumberOfSetsRemovedPerBlock {
+			return breakForEachPseudoError
+		}
+		if expiry := setExpiry.Get(id); expiry == nil {
+			setsWithoutExpiry = append(setsWithoutExpiry, id)
+			numSetsWithoutExpiry++
+		}
+
+		return nil
+	})
+	if err != nil && err.Error() != breakForEachPseudoError.Error() {
+		return err
+	}
+
+	if numSetsWithoutExpiry == 0 {
+		log.Info("Expiry information for all replay sets already " +
+			"known")
+		// Remember that there are no replay sets without expiry. As we
+		// store the expiry for new sets, we can stop scanning for old
+		// sets.
+		meta := tx.ReadWriteBucket(metaBucket)
+		err := meta.Put(allReplaySetsHaveExpiryKey, []byte{1})
+		if err != nil {
+			log.Errorf("Unable to update bucket meta "+
+				"data: %v", err)
+			return err
+		}
+
+		return nil
+	}
+
+	// If we found at least one replay set without expiry, we
+	// compute the maximum CLTV of all entries in sharedHashBucket
+	// and use this value to over-approximate the expiry of the
+	// replay sets without expiry.
+	maximumCltv, err := GetMaxCltvForSharedHashes(tx, err)
+	if err != nil {
+		return err
+	}
+	log.Infof("Setting expiry=%d for %d sets without expiry",
+		maximumCltv, numSetsWithoutExpiry)
+
+	// Set the expiry for each of the collected sets to the maximum
+	// CLTV so that the garbage collector will remove them
+	// eventually.
+	var scratch [4]byte
+	binary.BigEndian.PutUint32(scratch[:], maximumCltv)
+
+	setExpiryReadWrite := tx.ReadWriteBucket(replaySetExpiryBucket)
+	for _, id := range setsWithoutExpiry {
+		if err := setExpiryReadWrite.Put(id, scratch[:]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// AllReplaySetsHaveExpiry returns true only if for all entries in
+// batchReplayBucket we have a corresponding expiry entry in
+// replaySetExpiryBucket.
+func AllReplaySetsHaveExpiry(tx kvdb.RwTx) bool {
+	meta := tx.ReadBucket(metaBucket)
+	data := meta.Get(allReplaySetsHaveExpiryKey)
+	if data == nil {
+		return false
+	}
+
+	return data[0] == 1
+}
+
+// GetMaxCltvForSharedHashes iterates over all entries in sharedHashBucket and
+// returns the maximum CLTV value.
+func GetMaxCltvForSharedHashes(tx kvdb.RwTx, err error) (uint32, error) {
```

Comment:
I don't think this works in certain scenarios. If I send you an HTLC with a CLTV of max uint32 before you upgrade, you'll persist it in `sharedHashes` and then none of your batches will get gc'd

Reply:
In this case we wouldn't delete too much, which was my main concern. Do you think the current code (GC for `sharedHashes`) is unable to remove such entries? If so, we'd need to fix this issue first.

Reply:
`sharedHashes` doesn't have a maximum, it goes case-by-case so there's not really anything we can do about it there

Reply:
For the `sharedHashes` case, it's only one specific hash that couldn't be deleted so it's not a big deal. Also, somebody would actually have to modify their node to send super-high CLTV values. I only pointed it out because the way of choosing a maximum and then deleting in bulk may not work _at all_ if somebody were to actually do that